components/EditModal.tsx

```diff
@@ -0,0 +1,117 @@
+import {
+  Dispatch,
+  FormEvent,
+  SetStateAction,
+  useEffect,
+  useState,
+} from 'react';
+import { Gift } from '~/pages';
+import { Modal } from './Modal';
+import { TitleText } from './TitleText';
+import { Button } from './Button';
+import { updateGift } from '~/utils/jsonServerFunctions';
+import { Input } from './Input';
+import { SvgCheckMarkIcon } from '~/icons/CheckMarkIcon';
+import { SvgDeclineIcon } from '~/icons/DeclineIcon';
+
+type EditModal = {
+  gift: Gift;
+  refreshGiftList: () => void;
+  setIsModalOpen: Dispatch<SetStateAction<boolean>>;
+};
+
+export function EditModal({
+  gift,
+  refreshGiftList,
+  setIsModalOpen,
+}: EditModal) {
+  const [giftReceiver, setGiftReceiver] = useState('');
+  const [giftName, setGiftName] = useState('');
+
+  useEffect(() => {
+    setGiftName(gift.gift);
+    setGiftReceiver(gift.name);
+  }, [gift]);
+
+  useEffect(() => {
+    function handleKeyDown(e: KeyboardEvent) {
+      if (e.key === 'Escape') {
+        refreshGiftList();
+        setIsModalOpen(false);
+      }
+    }
+    document.addEventListener('keydown', handleKeyDown);
+    return function clearFunctions() {
+      document.removeEventListener('keydown', handleKeyDown);
+    };
+  }, [refreshGiftList, setIsModalOpen]);
+
+  async function handleEdit(e: FormEvent<HTMLElement>) {
+    e.preventDefault();
+
+    const newGift = gift;
```

Comment:
Tässä `newGift` ei ole kopio vaan viittaa samaan objektiin muistissa. Sen seurauksena seuraavalla rivillä muokkaat alkuperäistä objektia eli muokkaat funktion parametria mikä ei ole hyvä, [Täältä kannattaa lukea lisää](https://www.freecodecamp.org/news/javascript-assigning-values-vs-assigning-references/) value vs reference erosta, on tärkeä JavaScriptin käsite.

Reply:
Jos kopioitava objekti ei sisällä sisäkkäisiä objekteja, niin silloin teen kopion näin (sama tulos kuin sun Object.assignilla):
```ts
const newGift = { ...gift};
```
Jos taas kopioitavassa objektissa on sisäkkäisiä objekteja, niin ne eivät kopioidu tuolla. Silloin helpointa on käyttää `structuredClone` funktiota:
```ts
const newGift = structuredClone(gift);
```